README.md
bugfree
=======

BugFree API Node.js client.

* [BugFree API document](http://bugfree.corp.taobao.com/doc/openapi3.html)

## Install

```bash
$ npm install bugfree
```

## Usage

```js
var bugfree = require('bugfree');

var client = bugfree.create({
  api: 'http://your-bugfree-host/api3.php',
  apiKey: 'your api key'
});

client.users.getsid(function (err, session) {
  console.log(session);
});
```

## Authors

```bash
$ git summary

 project  : bugfree
 repo age : 2 minutes
 active   : 1 days
 commits  : 1
 files    : 12
 authors  :
     1  fengmk2                 100.0%
```

## License

(The MIT License)

Copyright (c) 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.0.1 / 2013-12-12 
==================

  * first commit


test/client.js
var bugfree = require('../');

var client = bugfree.create(require('./config'));

module.exports = client;

test/client.test.js
/*!
 * bugfree - test/client.test.js
 *
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var should = require('should');
var client = require('./client');

describe('client.test.js', function () {

  describe('getsid()', function () {
    it('should get session info', function (done) {
      client.getsid(function (err, session) {
        should.not.exist(err);
        session.should.have.keys('status', 'code', 'info', 'sessionname', 'sessionid', 'rand');
        session.status.should.equal('success');
        session.code.should.equal(0);
        session.sessionname.should.equal('PHPSESSID');
        session.sessionid.should.length(32);
        session.rand.should.length(5);
        done();
      });
    });
  });

  describe('login()', function () {
    it('should login success', function (done) {
      client.login(function (err, info) {
        should.not.exist(err);
        info.should.have.keys('status', 'info', 'code', 'timeout');
        client._session.timeout.should.equal(info.timeout * 1000 - 30000);
        info.timeout.should.above(Date.now() / 1000 + 3600 * 20);
        done();
      });
    });

    it('should login again success', function (done) {
      client.login(function (err, info) {
        should.not.exist(err);
        done();
      });
    });
  });

  // http://bugfree.corp.taobao.com/bug/257528
  var bugid = 257528;

  describe('getBug()', function () {
    it('should return a bug', function (done) {
      client.getBug(bugid, function (err, bug) {
        should.not.exist(err);
        // console.log(bug);
        bug.id.should.equal(String(bugid));
        done();
      });
    });
  });

  describe('updateBug()', function () {
    it('should update a bug status and add note', function (done) {
      var note = 'commit by @唐逸 at ' + Date() + ':<br><br>' +
        '这是一个 fixbug commit. <a href="http://gitlab.alibaba-inc.com/edp/alimovie/commit/97c1d509025d4fe7d211abbf94b2744589820bd4" target="_blank">97c1d5</a>';

      var data = {
        id: bugid,
        no_bbcode_transfer: 1,
        bug_status: 'Resolved',
        solution: 'see commit log',
        action_note: note,
        ResolvedBuild: '97c1d5',
      };
      client.updateBug(data, function (err, result) {
        should.not.exist(err);
        result.should.eql({ status: 'success', code: 0, info: '操作成功', BugID: '257528' });
        done();
      });
    });

    it('should relation a bug with note', function (done) {
      var note = 'commit by @唐逸 at ' + Date() + ':<br><br>' +
        '这是一个关联bug commit. <a href="http://gitlab.alibaba-inc.com/edp/alimovie/commit/97c1d509025d4fe7d211abbf94b2744589820bd4" target="_blank">97c1d5</a>';
      var data = {
        id: bugid,
        no_bbcode_transfer: 1,
        action_note: note,
      };
      client.updateBug(data, function (err, result) {
        should.not.exist(err);
        result.should.eql({ status: 'success', code: 0, info: '操作成功', BugID: '257528' });
        done();
      });
    });
  });

});


lib/client.js
/*!
 * bugfree - lib/bugfree.js
 *
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var debug = require('debug')('bugfree');
var urllib = require('urllib');
var utility = require('utility');
var eventproxy = require('eventproxy');

/**
 * Create a bugfree API client.
 *
 * @param {Object} options
 *  - {String} api, api root url, e.g.: 'http://your-bugfree-host/api3.php'
 *  - {String} apiKey
 *  - {String} username
 *  - {String} password
 */
function Client(options) {
  this.api = options.api;
  this.apiKey = options.apiKey;
  this.username = options.username;
  this.password = options.password;
  this._session = null;
}

Client.prototype.getBug = function (id, callback) {
  this.request('get', 'getbug', {id: id}, {}, callback);
};

Client.prototype.updateBug = function (data, callback) {
  this.request('post', 'updatebug', data, {}, callback);
};

Client.prototype.getsid = function (callback) {
  this.request('post', 'getsid', {}, {}, callback);
};

Client.prototype.login = function (callback) {
  if (this._session && this._session.timeout > Date.now()) {
    // has logined
    return callback();
  }

  var ep = eventproxy.create();
  ep.fail(callback);

  var that = this;
  that.getsid(ep.done('session'));

  ep.on('session', function (session) {
    var auth = utility.md5(utility.md5(that.username + utility.md5(that.password)) + that.apiKey + session.rand);
    var data = {
      username: that.username,
      auth: auth,
    };

    data[session.sessionname] = session.sessionid;
    that.request('post', 'login', data, {}, function (err, info, res) {
      if (err) {
        return callback(err);
      }
      if (info.status !== 'success') {
        err = new Error(that.username + ' login fail, ' + info.info);
        err.name = 'BugFreeLoginFailError';
        err.data = info;
        return callback(err);
      }
      session.timeout = info.timeout * 1000 - 30000; // before 30 seconds check
      that._session = session;
      callback(null, info);
    });
  });
};

Client.prototype.request = function (method, mode, data, options, callback) {
  var that = this;
  var url = that.api + '?mode=' + mode;
  data = data || {};
  if (that._session && !data[that._session.sessionname]) {
    // auto add session id
    data[that._session.sessionname] = that._session.sessionid;
  }
  options = options || {};
  options.timeout = options.timeout || that.requestTimeout;
  options.method = method || 'POST';
  options.dataType = options.dataType || 'json';
  options.data = data;
  urllib.request(url, options, function (err, resData, res) {
    debug('%s %s %j: status: %s, resData: %d bytes, err: %j',
      method, url, data, res && res.statusCode, resData && resData.length || 0, err);

    that.handleResult(err, resData, res, function (err, result) {
      if (err) {
        err.url = url;
        err.method = method;
      }
      callback(err, result, res);
    });
  });
};

Client.prototype.handleResult = function (err, result, res, callback) {
  var statusCode = res && res.statusCode;
  var headers = res && res.headers || {};
  if (err) {
    if (err.name === 'SyntaxError') {
      err.name = this.constructor.name + 'ReponseFormatError';
      if (res) {
        err.message = 'Parse ' + this.contentType + ' error: ' + err.message;
      }
    } else {
      err.name = this.constructor.name + err.name;
    }
  } else if (statusCode > 300) {
    var errorInfo = result || {};
    err = new Error(errorInfo.message ? errorInfo.message : 'Unknow Error ' + statusCode);
    if (errorInfo.name) {
      err.name = this.constructor.name + errorInfo.name;
    } else {
      err.name = this.constructor.name + statusCode + 'Error';
    }
    err.errors = errorInfo.errors;
  }

  if (err) {
    err.headers = headers;
    if (Buffer.isBuffer(result)) {
      result = result.toString();
    }
    err.data = { resBody: result };
    err.statusCode = statusCode;
    result = null;
  }

  callback(err, result);
};

exports.create = function (options) {
  return new Client(options);
};
